src/libsyntax/parse/parser/diagnostics.rs

```diff
@@ -1140,17 +1140,24 @@ impl<'a> Parser<'a> {
                     delim.to_string(),
                     Applicability::MaybeIncorrect,
                 );
-                err.emit();
-                self.expected_tokens.clear();  // reduce errors
-                Ok(true)
+                if unmatched.found_delim.is_none() {
+                    // Encountered `Eof` when lexing blocks, do not recover here to avoid knockdown
+                    // errors which will be emitted upstream in the parser and let other error
+                    // recovery consume the rest of the file.
+                    Err(err)
+                } else {
+                    err.emit();
+                    self.expected_tokens.clear();  // reduce errors
+                    Ok(true)
+                }
             }
             _ => Err(err),
         }
     }
 
     /// Recovers from `pub` keyword in places where it seems _reasonable_ but isn't valid.
     pub(super) fn eat_bad_pub(&mut self) {
-        if self.token.is_keyword(kw::Pub) {
+        if self.token.is_keyword(kw::Pub) && self.unclosed_delims.is_empty() {
```

Comment:
`&& rhs` wants a comment, mostly for the benefit of future me as this function might not exist anymore once I've fixed https://github.com/rust-lang/rust/issues/65041.